src/Simplify_Call.cpp

```diff
@@ -125,56 +125,40 @@ Expr Simplify::visit(const Call *op, ExprInfo *bounds) {
             return a;
         }
 
+        const bool shift_left = op->is_intrinsic(Call::shift_left);
+        const Type t = op->type;
+
         int64_t ib = 0;
-        if (const_int(b, &ib) || const_uint(b, (uint64_t *)(&ib))) {
-            Type t = op->type;
-
-            bool shift_left = op->is_intrinsic(Call::shift_left);
-            if (ib < 0) {
-                constexpr int64_t i64_max = std::numeric_limits<int64_t>::max();
-                if (t.is_int()) {
-                    shift_left = !shift_left;
-                    // Ensure that corner case of ib = lowest() doesn't negate into itself
-                    ib = -std::max(ib, -i64_max);
+        uint64_t ub = 0;
+        if (const_int(b, &ib)) {
+            // LLVM shl and shr instructions produce poison for negative shifts,
+            // or for shifts >= typesize, so we will follow suit in our simplifier.
+            user_assert(ib >= 0) << "bitshift by a constant negative amount is not legal in Halide.";
```

Comment:
I think this should possibly simplify to signed_integer_overflow instead of throwing an error. If we want to throw a user error, it should be in `Expr::operator<<(int)`

Reply:
See above re: immediate error. 

Re: immediate errors in `operator<<`` and friends, I considered doing this, but it wasn't clear to me how much precedent there was for this sort of thing. (I am a fan of failing hard-and-early for this sort of thing, though.)